node/src/components/consensus/highway_core/highway.rs

```diff
@@ -518,16 +524,24 @@ impl<C: Context> Highway<C> {
         evidence: Evidence<C>,
         rng: &mut dyn CryptoRngCore,
     ) -> Vec<Effect<C>> {
-        let was_honest = !self.state.is_faulty(evidence.perpetrator());
-        self.state.add_evidence(evidence.clone());
-        let mut effects = self.on_new_evidence(&evidence, rng);
-        if was_honest {
+        if self.state.add_evidence(evidence.clone()) {
+            let mut effects = self.on_new_evidence(&evidence, rng);
+            // Add newly created endorsements to the local state.
+            for effect in effects.iter() {
+                if let Effect::NewVertex(vv) = effect {
+                    if let Some(e) = vv.endorsements() {
+                        self.state.add_endorsements(e.clone());
```

Comment:
Wouldn't it be easier to add them directly in `on_new_evidence`?
Whenever we destructure `Effect`s inside `highway_core` it feels like we're doing something in the wrong order…

Reply:
I agree but I will have to destructure it anyway since `active_validator.on_new_evidence` returns `Vec<Effect<C>>`.

Reply:
I've rearranged the code a bit, although this destructuring had to stay.